pkg/objstore/oss/oss.go

```diff
@@ -0,0 +1,310 @@
+	for {
+		select {
+		case <-ctx.Done():
+			return ctx.Err()
+		default:
+		}
+		objects, err2 := b.bucket.ListObjects(alioss.Prefix(dir), alioss.Delimiter(objstore.DirDelim), marker)
+		if err2 != nil {
+			return errors.Wrap(err2, "oss bucket list")
+		}
+		for _, object := range objects.Objects {
+
+			if object.Key == "" {
+				continue
+			}
+			if err := f(object.Key); err != nil {
+				return err
+			}
+		}
+		for _, object := range objects.CommonPrefixes {
+			if err := f(object); err != nil {
+				return err
+			}
+		}
+		marker = alioss.Marker(objects.NextMarker)
+		if !objects.IsTruncated {
+			break
+		}
+	}
+
+	return nil
+}
+
+func (b *Bucket) Name() string {
+	return b.name
+}
+
+// validate checks to see the config options are set.
+func validate(conf Config) error {
+	if conf.EndPoint == "" {
+		return errors.New("no oss endpoint in config file")
+	}
+	if conf.Bucket == "" {
+		return errors.New("no oss bucket in config file")
+	}
+
+	if conf.SecretId == "" && conf.SecretKey != "" {
+		return errors.New("no oss acccess_key specified while secret_key is present in config file; both of them should be present in config.")
+	}
+
+	if conf.SecretId != "" && conf.SecretKey == "" {
+		return errors.New("no oss secret_key specified while access_key is present in config file; both of them should be present in config.")
+	}
+	return nil
+}
+
+func NewTestBucketFromConfig(t testing.TB, location string, c Config, reuseBucket bool) (objstore.Bucket, func(), error) {
+	bc, err := yaml.Marshal(c)
+	if err != nil {
+		return nil, nil, err
+	}
+	b, err := NewBucket(log.NewNopLogger(), bc, "thanos-e2e-test")
+	if err != nil {
+		return nil, nil, err
+	}
+
+	bktToCreate := c.Bucket
+	if c.Bucket != "" && reuseBucket {
+		if err := b.Iter(context.Background(), "", func(f string) error {
+			return errors.Errorf("bucket %s is not empty", c.Bucket)
+		}); err != nil {
+			return nil, nil, errors.Wrapf(err, "oss check bucket %s", c.Bucket)
+		}
+
+		t.Log("WARNING. Reusing", c.Bucket, "Aliyun OSS bucket for OSS tests. Manual cleanup afterwards is required")
+		return b, func() {}, nil
+	}
+
+	if c.Bucket == "" {
+		src := rand.NewSource(time.Now().UnixNano())
+
+		bktToCreate = strings.Replace(fmt.Sprintf("test_%s_%x", strings.ToLower(t.Name()), src.Int63()), "_", "-", -1)
+		if len(bktToCreate) >= 63 {
+			bktToCreate = bktToCreate[:63]
+		}
+	}
+
+	if err := b.client.CreateBucket(bktToCreate, alioss.ACL(alioss.ACLPrivate), alioss.StorageClass(alioss.StorageArchive)); err != nil {
+		return nil, nil, err
+	}
+	b.name = bktToCreate
+	
+	return b, func() {
+		objstore.EmptyBucket(t, context.Background(), b)
+		if err := b.client.DeleteBucket(bktToCreate); err != nil {
+			t.Logf("deleting bucket %s failed: %s", bktToCreate, err)
+		}
+	}, nil
+}
+
+func (b *Bucket) Close() error { return nil }
+
+func setRange(start, end int64) (alioss.Option, error) {
+	var opt alioss.Option
+	if 0 <= start && start <= end {
+		opt = alioss.Range(start, end)
```

Comment:
End can be negative. For example, [100, -1] means return all exception first 100 bytes.

Reply:
@wujinhu , It is not possible ,  see getRange method  , parameter is off(set) & length ,  so the `end` can not be negative

Reply:
Right, got it.